src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Collection/JsonCollectionConverter.cs

```diff
@@ -304,7 +312,9 @@ protected static JsonConverter<TElement> GetElementConverter(ref WriteStack stat
 
                     // Writing the start of the array must happen after any metadata
                     writer.WriteStartArray();
-                    state.Current.JsonPropertyInfo = state.Current.JsonTypeInfo.ElementTypeInfo!.PropertyInfoForTypeInfo;
+                    state.Current.JsonPropertyInfo = jsonTypeInfo.ElementTypeInfo!.PropertyInfoForTypeInfo;
+
+                    jsonTypeInfo.OnSerializing?.Invoke(value);
```

Comment:
Nit: I would move the callback before the first JSON gets written to the underlying writer.

Reply:
I agree, it was my first guess, but the `ObjectDefaultConverter` class triggers the callback after metadata and writing the StartObject token.
Could it be an issue?

Reply:
Done